packages/grid/x-data-grid/src/internals/public.ts

```diff
@@ -0,0 +1,12 @@
+// Only export the variable and types that should be publicly exposed and re-exported from `@mui/x-data-grid-pro`
+export * from './components/public';
```

Comment:
I can revert the logic yes

Reply:
@m4theushw small point to consider here `packages/x-data-grid/src/internals/index.ts` must be the private version 
So that we can import it from `@mui/x-data-grid/internals`

If we switch the other `index.ts` to be the public exports, it would be a small inconsistency.

Do you still want to do the switch ?

Reply:
It confuses me because we have a `public.ts` file inside an `internals` folder. We import from `internals` even things already public. I'm favor of using `index.ts` for public exports because it's how JavaScript modules work. Based on that, I have the following proposal:

1. Move everything inside `packages/grid/x-data-grid/src/internals/` one level up. `packages/grid/x-data-grid/src/internals/gridClasses.ts` becomes `packages/grid/x-data-grid/src/gridClasses.ts`.
2. Re-exports the public modules in `packages/grid/x-data-grid/src/index.ts`
3. Inside `packages/grid/x-data-grid/src/internals/index` we export only the true internal modules.

> I prefer the public, because filename index.ts is common, whereas with public.ts I would think "Be careful, this is the public export, are you sure it is the interface you want to maintain?"

In the past we had the bad habit of including everything inside `index.ts`. Because of that, many things were exported by mistake. Now we tend to include in it only things used in a demo or that users will certainly want. I think with the export snapshot we don't have this problem anymore.

Reply:
But how do you handle deep colocated elements ?
For instance in `packages/grid/x-data-grid/src/internals/hooks/features/columns/gridColumnsInterfaces.ts` we have both public and private interfaces.
And splitting those files into two distincts folders seems very complicated to me.

Making `index.ts` public and `private.ts` private is fine for me.
But making `@mui/x-data-grid/internals` not containing the public stuff is a lot of work and I don't see the point?

Reply:
> But how do you handle deep colocated elements ?

By individually exporting the interfaces, instead of usuing `export * from`. We need to enable `isolatedModules` for that.

```diff
diff --git a/packages/grid/x-data-grid/src/internals/hooks/features/columns/index.ts b/packages/grid/x-data-grid/src/internals/hooks/features/columns/index.ts
index 524807ffb..6d4f470ce 100644
--- a/packages/grid/x-data-grid/src/internals/hooks/features/columns/index.ts
+++ b/packages/grid/x-data-grid/src/internals/hooks/features/columns/index.ts
@@ -1,4 +1,9 @@
 export * from './gridColumnsSelector';
-export * from './gridColumnsInterfaces';
+export {
+  GridColumnLookup,
+  GridColumnsState,
+  GridColumnsInitialState,
+  GridColumnVisibilityModel,
+} from './gridColumnsInterfaces';
 export * from './gridColumnsUtils';
 export * from './useGridColumns';
```

Keeping a file only for public/private exports works but I think it requires more mental effort to ignore the `index.ts` when we need to think if something will be included in the public API or not. I initially proposed to rename to `private.ts` but maybe this will only avoid a little bit the confusion.

Reply:
>  We need to enable isolatedModules for that.

Don't think we need it, we are already exporting it like that on some files.

But that's not my main concern.
It more that I don't get the folder architecture you want.
In which file would be `GridColumnsRawState` for instance ? It's private, must be accessible from `@mui/x-data-grid-pro`, but highly linked to public elements that are currently in the same file.

Would you put it in an entirely different file, in `src/internals/....` (whereas the publci would be in `src/hooks/...`.
Or would you have `src/internals/index.ts` export directly from the deep file where is curently is ?


Note that `src/internals/index.ts` is needed instead of `src/internals.ts` for bundling reason. Even if it's the only file of its directory.

Reply:
> It more that I don't get the folder architecture you want.

```
package.json
src/DataGrid.tsx
src/index <- Exports to be available under @mui/x-data-grid
src/hooks/features/columns/index.ts
src/hooks/features/columns/gridColumnsInterfaces.ts <- GridColumnsRawState is still here
src/internals/index.ts <- Exports to be available under @mui/x-data-grid/internals
```

Contents of `src/hooks/features/columns/index.ts`:

```ts
export * from './gridColumnsSelector';
export * from './gridColumnsInterfaces';
export {
  GridColumnLookup,
  GridColumnsState,
  GridColumnsInitialState,
  GridColumnVisibilityModel,
} from './gridColumnsInterfaces';
export * from './gridColumnsUtils';
export * from './useGridColumns';
```

Contents of `src/internals/index.ts`:

```tsx
exports { GridColumnsRawState } from '../hooks/features/columns/gridColumnsInterfaces';
```

This resembles what we had before the introduction of the `internals` folder. Also, forces us to import true internal things from a different import in DataGridPro.

Reply:
I can't have files at the root of `src`
Otherwise, the core bundling strategy don't work correctly and I get bundle duplication (and React context problems)

So to make things work, I would need to do:

```
package.json
src/DataGrid/DataGrid.tsx <- The folder would also contains `useDataGridProps.ts` and `useDataGridComponent.ts`
src/index.ts <- Exports to be available under @mui/x-data-grid
src/hooks/features/columns/index.ts
src/hooks/features/columns/gridColumnsInterfaces.ts <- GridColumnsRawState is still here
src/internals/index.ts <- Exports to be available under @mui/x-data-grid/internals
```

Reply:
> I can't have files at the root of src

Which type of file can't be at the root of `src`? Your example contains files at this location.

If the structure you proposed can be done I would agree to adopt it. It's flatter than having an `internals` folder wrapping everything. The core bundling process can be improved later to avoid to create a folder for `DataGrid.tsx`.

Reply:
Any file not named `index.ts`
It is the `topLevelPathImportsCanBePackages` condition on the bundling process.
The condition make sense.